Display status emoji in several additional places in the UI

YashRE42 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This continues the work from #20019. In order to address #19865.

**Testing plan:** <!-- How have you tested? -->
automated tests pass, and manual testing on chrome/windows did not reveal any bugs.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
<details>
<summary>
working GIF
</summary>

![](https://user-images.githubusercontent.com/33805964/152673045-513f16c7-d1da-4291-a01c-59e868ed90a4.gif)
</details>

<details>
<summary>
Live updates for status emoji or user renames do not work on compose_pm pills
</summary>

![](https://user-images.githubusercontent.com/33805964/152673042-d2563d42-cde2-41be-8e0f-3c7582658e88.gif)
</details>

timabbott commented on static/styles/right_sidebar.css:
```diff
@@ -113,14 +113,7 @@
-        margin-left: 4px;
-        margin-right: 4px;
-        top: 8px;
+        top: 9px;
```
Why the 8px => 9px change?

YashRE42 replied:
I'm not exactly sure _why_ it has to be different here, but this is pixel perfect (ie the same as other places) whereas 8px would be 1px off.

timabbott replied:
OK.  Also going to move this to `app_components.css`; we put reusable CSS there, not in `zulip.css` which we're trying to slowly kill.

timabbott replied:
Hmm, maybe worth doing later, since we want to move `emoji` at the same time.

timabbott commented on static/styles/user_status.css:
```diff
@@ -29,9 +29,12 @@
+                    the following rule is not necessarily better than the one
+                    from a6bef5154197b15c20445526fd3f219b4f2e5379 but it works.
+                */
+                top: -2px;
```
Does this interact well with zooming the browser window?  I'd generally expect that to be an advantage of explicit vertical centering styling.  like what was there before.  (Though I feel like we might be able to use more modern CS alignment things to just say "please center this vertically")

timabbott replied:
Also I don't quite understand under what circumstances this CSS is used -- can you explain?

timabbott replied:
OK, I'm not sure why the -2px should be correct, but I guess this works.

timabbott commented on frontend_tests/node_tests/pm_list.js:
```diff
@@ -94,6 +105,7 @@ test("build_private_messages_list", ({override}) => {
             is_active: false,
             url: "#narrow/pm-with/101,102-group",
             user_circle_class: undefined,
+            status_emoji_info: undefined,
```
Do we want a test where `status_emoji_info` is not undefined to exercise the new code?

timabbott commented on static/templates/status_emoji.hbs:
```diff
@@ -0,0 +1,13 @@
+        <div class="emoji status_emoji emoji-{{status_emoji_info.emoji_code}}"></div>
+        {{/if}}
+    {{/if}}
+{{/if}}
```
I would usually expect this sort of template extraction to be passed `status_emoji_info` from the caller / calling template, and just access fields like `emoji_code` here.

YashRE42 replied:
Done.

Also noticed that we were inserting `data-still-url="{{still_url}}"` in the case where `{{#if still_url}}` was false so stopped doing that.

timabbott commented on static/js/message_live_update.js:
```diff
@@ -28,5 +28,11 @@ export function update_avatar(user_id, avatar_url) {
+export function update_user_status_emoji(user_id, status_emoji_info) {
+    message_store.update_property("status_emoji_info", status_emoji_info, {user_id});
     rerender_messages_view();
 }
```
I think this implementation, as written, will trigger a full rerender of the displayed message list every time someone changes their status, regardless of whether any messages sent by that user were visible.

This was OK previously because fields like name very rarely change.  But we probably will want to make the rerender conditional on the user whose status was changed actually having any messages that they'd sent.

Let's split fixing this out as a small preparatory PR that changes this interface here with the existing "change name" code path -- I think basically we want to replace `rerender_messages_view` function in this module to instead be `rerender_messages_view_if_has_sender(user_id)`, and then to add a conditional that calls into a `message_list_data.js` function.  You can probably use this as inspiration:
```
    get_last_message_sent_by_me() {                                                                      
        const msg_index = _.findLastIndex(this._items, {sender_id: page_params.user_id});                
        if (msg_index === -1) {                                                                          
            return undefined;                                                                            
        }                                                                                                
        const msg = this._items[msg_index];                                                              
        return msg;                                                                                      
    }                                                                                                    
```

YashRE42 replied:
(working on it)

YashRE42 replied:
opened #21081

timabbott commented on static/js/settings_user_groups.js:
```diff
@@ -72,7 +77,16 @@ export function populate_user_groups() {
+            // containers and break.) An alternative approach would be to use a
+            // function generator, but the best solution would probably be to
+            // redo the pills system using modern JavaScript classes. TODO?
+            user_pill.append_user.call(user_pill_config, user, pills);
```
This is pretty ugly and I think we want a better solution.  Why can't we just add `user_pill_config` as a parameter on the existing pill container objects when they are initialized?

YashRE42 replied:
Right, I was trying to prevent `emoji_status_info` from being passed down to input_pills so that it's not rendered into the template. Except it's simpler to just add `pill_config` to the `input_pill` `store` and check for that in `input_pill`.

I've updated to that approach.

timabbott replied:
lgtm.

timabbott commented:
I merged the first two commits as 5e05dcbc8f40c499ff420ddf815ff6029e5215e4.  Definitely looking promising!

YashRE42 commented:
I forgot to comment yesterday when I pushed to this, I added the test we were looking for.

In the last push, I also addressed https://chat.zulip.org/#narrow/stream/101-design/topic/moved.20indicator/near/1325213.
![](https://user-images.githubusercontent.com/33805964/153651843-75e200d2-d3a6-48bf-ad99-990c59f99f78.png)

and another piece of feedback from @alya
> We should show statuses in group PMs in the left sidebar ...

![](https://user-images.githubusercontent.com/33805964/153652488-9dd5b21d-affd-4ff6-870a-5e26c1fc27c4.png)

<details>
<summary>
live update for group pms
</summary>

![](https://user-images.githubusercontent.com/33805964/153653621-4ed882c9-e47e-4e4d-a1d8-f25883fe7128.gif)
</details>

andersk commented on static/js/people.js:
```diff
@@ -306,20 +306,22 @@ export function get_full_names_for_poll_option(user_ids) {
-            blueslip.error("Unknown user id " + user_id);
-            return "?";
-        }
+function map_user_id_to_name(user_id) {
```
This function doesn’t map anything. It’s just `get_display_full_name`.

YashRE42 replied:
renamed to `get_display_full_name`, thanks!

timabbott commented:
I merged most of this as the series ending with 96682a351707e614600c829203f9a2680a9f5c50, and pushed the bits related to just the "Group PMs" region back here.  Thanks for doing this @YashRE42!  Here's the total diff for my changes to this PR.  As you can see, most of the changes are deleting inconsistent spacing CSS that I didn't find improved the UI (Alya did a shoulder check and agreed).

```
diff --git a/static/styles/input_pill.css b/static/styles/input_pill.css
index 8718785884..119041eb19 100644
--- a/static/styles/input_pill.css
+++ b/static/styles/input_pill.css
@@ -110,10 +110,6 @@
         content: attr(data-some-recipients-text);
         opacity: 0.5;
     }
-
-    .status_emoji {
-        margin-left: 0;
-    }
 }
 
 .deactivated-pill {
diff --git a/static/styles/zulip.css b/static/styles/zulip.css
index 281d4178c5..d3f1cddc79 100644
--- a/static/styles/zulip.css
+++ b/static/styles/zulip.css
@@ -1293,11 +1293,6 @@ td.pointer {
     font-weight: 700;
     vertical-align: top;
     line-height: 12px;
-
-    .status_emoji {
-        margin-left: 7px;
-        margin-right: 0;
-    }
 }
 
 .sender_name-in-status {
@@ -2465,8 +2460,13 @@ div.topic_edit_spinner .loading_indicator_spinner {
     /* We are setting minimum width here because when the user's name is very long,
     emoji's width decreases and causes it to break. */
     min-width: 16px;
-    margin-left: 4px;
-    margin-right: 4px;
+    /* In most contexts, status emoji appear immediately after a name
+      field with no margin. Position the status emoji with 2px of left
+      margin to space it from the name, and set no right margin so
+      that any components to the right appear equally distant as they
+      would be from a name. */
+    margin-left: 2px;
+    margin-right: 0;
 }
 
 /* FIXME: Combine this rule with the one in portico.css somehow? */
diff --git a/static/templates/input_pill.hbs b/static/templates/input_pill.hbs
index 371d77641d..429735c9e2 100644
--- a/static/templates/input_pill.hbs
+++ b/static/templates/input_pill.hbs
@@ -2,10 +2,11 @@
     {{#if has_image}}
     <img class="pill-image" src="{{img_src}}" />
     {{/if}}
-    <span class="pill-value">{{ display_value }}</span>
-    {{#if has_status}}
-    {{> status_emoji status_emoji_info}}
-    {{/if}}
+    <span class="pill-value">{{ display_value }}
+    {{~#if has_status~}}
+    {{~> status_emoji status_emoji_info~}}
+    {{~/if~}}
+    </span>
     <div class="exit">
         <span aria-hidden="true">&times;</span>
     </div>
diff --git a/static/templates/pm_list_item.hbs b/static/templates/pm_list_item.hbs
index 62ade522dd..146c82d105 100644
--- a/static/templates/pm_list_item.hbs
+++ b/static/templates/pm_list_item.hbs
@@ -10,9 +10,9 @@
         </div>
 
         <a href='{{url}}' class="conversation-partners">
-            {{#each recipients}}
-                {{#unless @first}}&nbsp;{{/unless}}{{> pm_list_item_recipient}}{{#unless @last}},{{/unless}}
-            {{/each}}
+            {{~#each recipients~}}
+                {{~#unless @first}}&nbsp;{{/unless}}{{~> pm_list_item_recipient~}}{{#unless @last}},{{/unless~}}
+            {{~/each~}}
         </a>
         <span class="unread_count {{#if is_zero}}zero_count{{/if}}">
             {{unread}}
```

The "Group PMs" changes feel too messy for me to want to integrate as of yet.  I think adding a `user_display.js` module rather than modifying `people.js` would be a good path forward here; also I think you should extract the logic to make `recipients` an object as its own prep commit that we can merge first; that'd declutter this a lot.

Now that this is merged, it's important that we push to

YashRE42 commented:
This last push addressed [Tim's comment on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Status.20emoji.20in.20message.20feed.20and.20PMs/near/1327227):
> Hmm, I guess we need to adjust margins for the sidebar applications, possibly left-margins on the unread counts are the right solution. I think logically we want the preserve the 1-sided margins on status emoji (possibly bumping from 2px to 3-4px as discussed in the last few messages), and just make sure the elements to their right have appropriate spacing.

But I have not yet made the changes for the group PMs

YashRE42 commented:
Last push was just to remove a merge conflict, group_pms changes are still not fixed.

Before / after for `right_sidebar: Apply left margin to unread counts.` [693c4e9](https://github.com/zulip/zulip/pull/21050/commits/693c4e998e94720210fc832c09b97bd025b7e175)
![](https://user-images.githubusercontent.com/33805964/156770868-f369fca7-6009-48fd-9a38-9ac357442652.png)

alya commented:
Hm, the "after" definitely looks better, but the octopus still seems a bit tight on the right to me. Is there less margin than on the lines without a status emoji?

YashRE42 commented:
Is this any better? bumped from 3px to 4px. (The margin is the same for both, but the browser decides where to start truncating from.)
![](https://user-images.githubusercontent.com/33805964/156912383-25c9f2d7-eb24-4343-9cb7-868d42bdaceb.png)

I also bumped the margin left for status emoji on these by one pixel, I somehow missed out on pushing this change earlier.
![](https://user-images.githubusercontent.com/33805964/156912447-2188531b-31c9-4483-a471-db614acf938b.png)

alya commented:
The updated screenshot of the left sidebar looks better to me!

alya commented:
The message feed one looks good as long as it's consistent with all the other places.

timabbott commented:
I merged a couple more preparatory pure refactoring commits from this PR while I was looking at it.

timabbott commented:
Closing as I think this was completed in https://github.com/zulip/zulip/commit/96682a351707e614600c829203f9a2680a9f5c50.
